unsafe-fields/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+impl<O: ?Sized, F: Copy, const NAME_HASH: u128> Clone for Unsafe<O, F, { NAME_HASH }> {
+    #[allow(clippy::non_canonical_clone_impl)]
+    fn clone(&self) -> Self {
+        // SAFETY: We don't call any user-defined code here (only make a
+        // bit-for-bit copy of `self.field`), so there's no way to accidentally
+        // invoke user-defined code or modify `self.field`.
+        Unsafe { _marker: PhantomData, field: self.field }
+    }
+}
+
+impl<O: ?Sized, F: ?Sized, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
+    /// Gets a reference to the inner value.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn as_ref(&self) -> &F {
+        // SAFETY: This method is unsafe to call.
+        &self.field
+    }
+
+    /// Gets a mutable reference to the inner value.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub unsafe fn as_mut(&mut self) -> &mut F {
+        // SAFETY: This method is unsafe to call.
+        &mut self.field
+    }
+}
+
+impl<O: ?Sized, F, const NAME_HASH: u128> Unsafe<O, F, { NAME_HASH }> {
+    /// Constructs a new `Unsafe<O, F, NAME_HASH>`.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn new(field: F) -> Unsafe<O, F, { NAME_HASH }> {
+        // SAFETY: This method is unsafe to call.
+        Unsafe { _marker: PhantomData, field }
+    }
+
+    /// Extracts the inner `F` from `self`.
+    ///
+    /// # Safety
+    ///
+    /// The caller is responsible for upholding any safety invariants associated
+    /// with this field.
+    #[inline(always)]
+    pub const unsafe fn into(self) -> F {
+        use core::mem::ManuallyDrop;
+
+        let slf = ManuallyDrop::new(self);
+
+        #[repr(C)]
+        union Transmute<O: ?Sized, F, const NAME_HASH: u128> {
+            src: ManuallyDrop<Unsafe<O, F, { NAME_HASH }>>,
+            dst: ManuallyDrop<F>,
+        }
+
+        // SAFETY: `ManuallyDrop<Unsafe<_, F, _>>` has the same size and bit
+        // validity as `Unsafe<_, F, _>`. [1] `Unsafe<_, F, _>` is
+        // `#[repr(transparent)]` and has no other fields, and so it has the
+        // same size and bit validity as `F`.
+        //
+        // [1] Per https://doc.rust-lang.org/1.81.0/core/mem/struct.ManuallyDrop.html:
+        //
+        //   `ManuallyDrop<T>` is guaranteed to have the same layout and bit
+        //   validity as `T`
+        let dst = unsafe { Transmute { src: slf }.dst };
+
+        // SAFETY (satisfaction of `Unsafe`'s field invariant): This method is
+        // unsafe to call.
+        ManuallyDrop::into_inner(dst)
```

Comment:
Could you just add a comment explaining why we need to go through all of this ceremony?

Reply:
Done.